flink-end-to-end-tests/flink-end-to-end-tests-common/src/main/java/org/apache/flink/tests/util/flink/container/FlinkContainers.java

```diff
@@ -0,0 +1,309 @@
+            Configuration conf,
+            Runnable cleanupHook) {
+        this.jobManager = jobManager;
+        this.taskManagers = taskManagers;
+        this.haService = haService;
+        this.conf = conf;
+        this.cleanupHook = cleanupHook;
+    }
+
+    /** Starts all containers. */
+    public void start() throws Exception {
+        if (haService != null) {
+            LOG.debug("Starting HA service container");
+            this.haService.start();
+        }
+        LOG.debug("Starting JobManager container");
+        this.jobManager.start();
+        waitUntilJobManagerRESTReachable(jobManager);
+        LOG.debug("Starting TaskManager containers");
+        this.taskManagers.parallelStream().forEach(GenericContainer::start);
+        LOG.debug("Creating REST cluster client");
+        this.restClusterClient = createClusterClient();
+        waitUntilAllTaskManagerConnected();
+        isStarted = true;
+    }
+
+    /** Stops all containers. */
+    public void stop() {
+        isStarted = false;
+        if (restClusterClient != null) {
+            restClusterClient.close();
+        }
+        this.taskManagers.forEach(GenericContainer::stop);
+        this.jobManager.stop();
+        if (this.haService != null) {
+            this.haService.stop();
+        }
+        cleanupHook.run();
+    }
+
+    /** Gets the running state of the cluster. */
+    public boolean isStarted() {
+        return isStarted;
+    }
+
+    /** Gets JobManager container. */
+    public GenericContainer<?> getJobManager() {
+        return this.jobManager;
+    }
+
+    /** Gets JobManager's hostname on the host machine. */
+    public String getJobManagerHost() {
+        return jobManager.getHost();
+    }
+
+    /** Gets JobManager's port on the host machine. */
+    public int getJobManagerPort() {
+        return jobManager.getMappedPort(this.conf.get(RestOptions.PORT));
+    }
+
+    /** Gets REST client connected to JobManager. */
+    public RestClusterClient<StandaloneClusterId> getRestClusterClient() {
+        return this.restClusterClient;
+    }
+
+    /**
+     * Restarts JobManager container.
+     *
+     * <p>Note that the REST port will be changed because the new JM container will be mapped to
+     * another random port. Please make sure to get the REST cluster client again after this method
+     * is invoked.
+     */
+    public void restartJobManager(RunnableWithException afterFailAction) throws Exception {
+        if (this.haService == null) {
+            LOG.warn(
+                    "Restarting JobManager without HA service. This might drop all your running jobs");
+        }
+        jobManager.stop();
+        afterFailAction.run();
+        jobManager.start();
+        // Recreate client because JobManager REST port might have been changed in new container
+        waitUntilJobManagerRESTReachable(jobManager);
+        this.restClusterClient = createClusterClient();
```

Comment:
AFAIK you need to close the old `restClusterClient` to prevent leaking resources.